src/utils/events/wheel-input.js

```diff
@@ -0,0 +1,133 @@
+    this.callback = callback;
+
+    this.handler = this.handler.bind(this);
+
+    WHEEL_EVENTS.forEach(eventName => element.addEventListener(eventName, this.handler));
+
+    this._state = {
+      mouseWheelPos: null
+    };
+  }
+
+  destroy() {
+    const {element} = this;
+    WHEEL_EVENTS.forEach(eventName => element.removeEventListener(eventName, this.handler));
+  }
+
+  /* eslint-disable complexity, max-statements */
+  handler(event) {
+
+    event.preventDefault();
+    let value = event.deltaY;
+    // Firefox doubles the values on retina screens...
+    if (firefox && event.deltaMode === window.WheelEvent.DOM_DELTA_PIXEL) {
+      value /= window.devicePixelRatio;
+    }
+    if (event.deltaMode === window.WheelEvent.DOM_DELTA_LINE) {
+      value *= WHEEL_DELTA_PER_LINE;
+    }
+
+    let type = this._state.mouseWheelType;
+    let timeout = this._state.mouseWheelTimeout;
+    let lastValue = this._state.mouseWheelLastValue;
+    let time = this._state.mouseWheelTime;
+
+    const now = (window.performance || Date).now();
+    const timeDelta = now - (time || 0);
+
+    const pos = {x: event.clientX, y: event.clientY};
+    time = now;
+
+    if (value !== 0 && value % WHEEL_DELTA_MAGIC_SCALER === 0) {
+      // This one is definitely a mouse wheel event.
+      type = 'wheel';
+      // Normalize this value to match trackpad.
+      value = Math.floor(value / WHEEL_DELTA_MAGIC_SCALER);
+    } else if (value !== 0 && Math.abs(value) < TRACKPAD_MAX_DELTA) {
+      // This one is definitely a trackpad event because it is so small.
+      type = 'trackpad';
+    } else if (timeDelta > 400) {
+      // This is likely a new scroll action.
+      type = null;
+      lastValue = value;
+      // Start a timeout in case this was a singular event, and delay it by up
+      // to 40ms.
+      timeout = window.setTimeout(function setTimeout() {
+        const _type = 'wheel';
+        this._wheel(event, -this._state.mouseWheelLastValue, this._state.mouseWheelPos);
+        this._setState({mouseWheelType: _type});
+      }.bind(this), 40);
+    } else if (!this._type) {
+      // This is a repeating event, but we don't know the type of event just
+      // yet.
+      // If the delta per time is small, we assume it's a fast trackpad;
+      // otherwise we switch into wheel mode.
+      type = Math.abs(timeDelta * value) < TRACKPAD_MAX_DELTA_PER_TIME ? 'trackpad' : 'wheel';
+
+      // Make sure our delayed event isn't fired again, because we accumulate
+      // the previous event (which was less than 40ms ago) into this event.
+      if (timeout) {
+        window.clearTimeout(timeout);
+        timeout = null;
+        value += lastValue;
+      }
+    }
+
+    if (event.shiftKey && value) {
+      value = value * SHIFT_MULTIPLIER;
+    }
+
+    // Only fire the callback if we actually know what type of scrolling device
+    // the user uses.
+    if (type) {
+      this._wheel(event, -value, pos);
+    }
+
+    this._setState({
```

Comment:
Nit: I actually think this `setState` function might not be needed and could even be a little confusing here. It doesn't add much to the a small class like this and for a moment makes me feel that I am reading a React component or a deck.gl layer. 

It is probably a matter of taste, but I personally try to carefully match the amount of "machinery" in a certain piece of code to be in proportion to the complexity of the code in question.